Added cosmic rate calculation tool

skeshri opened the pull request:
Added cosmic rate calculation tool for tracker alignment.

I have tested it locally before creating this pull request. Experts (@ghellwig and @mschrode) were monitoring this update.

skeshri committed:
Added cosmic rate calculation tool

cmsbuild commented:
A new Pull Request was created by  @skeshri  for CMSSW_8_1_X.

It involves the following packages:

Alignment/TrackerAlignment


@ghellwig, @cerminar, @cmsbuild, @franzoni, @mmusich, @davidlange6 can you please review it and eventually sign? Thanks.
@mschrode, @ghellwig, @mmusich, @tocheng, @tlampen this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

ghellwig commented on Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc:
```diff
@@ -0,0 +1,529 @@
+    treeRun = fs->make<TTree>("Run","");
+    treeCluster = fs->make<TTree>("Cluster","");
+
+}
```
@skeshri: you are using the TFileService, i.e. you have to add `usesResource(TFileService::kSharedResource);` in the constructor
in addition you need to add the template parameter `edm::one::SharedResources` to the `edm::one::EDAnalyzer` from which you inherit

ghellwig commented on Alignment/TrackerAlignment/plugins/BuildFile.xml:
```diff
@@ -35,3 +35,23 @@
 </library>
+
+<library file="CosmicRateAnalyzer.cc" name="AlignmentCosmicRateAnalyzerPlugins">
+  <use   name="Alignment/TrackerAlignment"/>
```
@skeshri this is not needed, I think. From the cc file I cannot see that you use something from this package.

ghellwig commented on Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc:
```diff
@@ -0,0 +1,529 @@
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void
+CosmicRateAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
```
@skeshri please add a suitable `fillDescriptions` method

ghellwig commented on Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py:
```diff
@@ -0,0 +1,39 @@
+			   )
+
+process.TFileService = cms.Service("TFileService", fileName = cms.string("Cosmic_rate_tuple.root") )
+process.demo = cms.EDAnalyzer("CosmicRateAnalyzer",
```
@skeshri when you have added the filldescriptions method you can load the auto-generated file here and modify only the parameters you want

ghellwig commented on Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py:
```diff
@@ -0,0 +1,39 @@
+#				tracks = cms.InputTag("ctfWithMaterialTracksP5"),                # Track collection for prompt RECO Dataset
+				tracks = cms.InputTag("ALCARECOTkAlCosmicsCTF0T"),		 # Track collection for stream and prompt ALCARECO Dataset
+				muons = cms.InputTag("muons1Leg"),		 # for muon Trigger timing information  
+#				fileName = cms.untracked.string("Cosmic_rate_tuple.root")
```
@skeshri remove this comment because you use the file service

ghellwig commented:
@skeshri please also link a presentation of your tool in the description of this pull request

cmsbuild commented:
+1
Tested at: 41b65cbc0611785def121172b49baf47f59cb951
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15150/13944/summary.html

skeshri committed:
added some missing features and cleaning up

cmsbuild commented:
Pull request #15150 was updated. @ghellwig, @cerminar, @cmsbuild, @franzoni, @mmusich, @davidlange6 can you please check and sign again.

mmusich commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1
@skeshri This pull request cannot be automatically  merged, could you please rebase it?

You can see the log for git cms-merge-topic here: 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15150/13959/git-merge-result

skeshri commented:
Hi I couldn't find any log at the link https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15150/13959/git-merge-result

mmusich commented:
@skeshri gitHub complains with: `This branch has conflicts that must be resolved`.
A rebase is likely needed

mmusich commented:
tip: try to merge your PR with `git cms-merge-topic 15150` in a fresh IB. That should show what went wrong.

skeshri commented:
I tried  git cms-merge-topic 15150. I got the following : 
From .
 * [new ref]                    -> merge-attempt
From https://github.com/cms-sw/cmssw
 * [new ref]         refs/pull/15150/head -> cms-sw/refs/pull/15150/head
Switched to branch 'merge-attempt'
Already up-to-date.
Switched to branch 'cosmic-rate-tool'
>> Checking Alignment/TrackerAlignment CMSSW_8_0_7_patch1
   x Alignment/TrackerAlignment/plugins/BuildFile.xml
Checking out these packages: 0

After that when I did git branch I got following : 
  CMSSW_8_0_X
  cms-sw/refs/pull/15150/head
* cosmic-rate-tool
  from-CMSSW_8_0_7_patch1

But I don't understand what exactly the problem is?

skeshri commented:
there is some problem in format.

these are four branches :
CMSSW_8_0_X
cms-sw/refs/pull/15150/head
cosmic-rate-tool 
from-CMSSW_8_0_7_patch1

mmusich commented:
@skeshri
```
cmsrel CMSSW_8_1_X_2016-07-10-2300
cd CMSSW_8_1_X_2016-07-10-2300/src/
cmsenv
git cms-merge-topic 15150
```

results in:
```
Auto-merging Alignment/TrackerAlignment/plugins/BuildFile.xml
CONFLICT (content): Merge conflict in Alignment/TrackerAlignment/plugins/BuildFile.xml
Automatic merge failed; fix conflicts and then commit the result.
Unable to merge branch refs/pull/15150/head from repository cms-sw.
```

you have an issue in: 

`Alignment/TrackerAlignment/plugins/BuildFile.xml`

skeshri commented:
@mmusich 
I have removed only one line from BuildFile.xml as it was not needed. You can see at  https://github.com/cms-sw/cmssw/pull/15150/commits/538cea4a3956530bfe1020672cfe74da38f7c072
But I couldn't trace the exact problem.

mmusich commented:
@skeshri try to rebase on top of the HEAD of CMSSW_8_1_X. It will fail the merge. You'll need to manually edit `Alignment/TrackerAlignment/plugins/BuildFile.xml` to remove the conflict symbols.
Once done commit and push again to the same branch.
Otherwise close this one and start from a fresh installation.

skeshri commented:
@mmusich 
Thanks Marco. I guess fresh installation is easier for me. I'm closing this pull request.
